compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp

```diff
@@ -872,42 +903,82 @@ FailureOr<SmallVector<Value>> Im2colOp::decomposeOperation(OpBuilder &b) {
-      std::min<int64_t>(getOutputRank(), getInputRank()), b.getIndexAttr(1));
-  kTileSizes.back() = kTileSize;
-
-  SmallVector<int64_t> kTileSizeStatic;
-  SmallVector<Value> kTileSizeDynamic;
-  dispatchIndexOpFoldResults(kTileSizes, kTileSizeDynamic, kTileSizeStatic);
-  auto extractType = cast<RankedTensorType>(outputType.clone(kTileSizeStatic));
+  int64_t inputRank = getInputRank();
+  int64_t outputRank = getOutputRank();
+
+  // For now, only extract 1d slice when the rank changes for vectorization
+  // along the batch dimension.
+  // TODO: check if 1d slice extraction can be used for all cases and thus
+  // simplify the codes.
+  SmallVector<OpFoldResult> inputTileSizes;
+  if (contiguousAlongB && inputRank != outputRank) {
+    inputTileSizes.push_back(innerInputTileSize);
+  } else {
+    int64_t minRank = std::min<int64_t>(inputRank, outputRank);
+    for (int i = 0; i < minRank - 1; i++) {
+      inputTileSizes.push_back(b.getIndexAttr(1));
+    }
+    inputTileSizes.push_back(innerInputTileSize);
+  }
+
+  SmallVector<int64_t> tileSizeStatic;
+  SmallVector<Value> tileSizeDynamic;
+  dispatchIndexOpFoldResults(inputTileSizes, tileSizeDynamic, tileSizeStatic);
+  auto extractType = cast<RankedTensorType>(outputType.clone(tileSizeStatic));
   auto extract =
       b.create<tensor::ExtractSliceOp>(nestedLoc, extractType, inputSlice,
                                        sliceOffsets, sliceSizes, sliceStrides);
 
   // Insert the slice into the destination tensor.
-  sliceOffsets = SmallVector<OpFoldResult>(getOutputRank(), zero);
-  sliceSizes = SmallVector<OpFoldResult>(getOutputRank(), one);
-  sliceStrides = SmallVector<OpFoldResult>(getOutputRank(), one);
-  sliceSizes.back() = kTileSize;
-  for (auto [idx, iv] : llvm::enumerate(ivs)) {
-    sliceOffsets[idx] = iv;
-  }
-  // Insert a `linalg.copy` so there is something to vectorize in the
-  // decomposition. Without this copy, the extract and insert slice ops
-  // do not get vectorized, and the sequence becomes a scalar memref.copy.
-  // This memref.copy could be vectorized after bufferization, but it is
-  // probably better to vectorize during generic vectorization.
-  Value copyDest = b.create<tensor::ExtractSliceOp>(
-      nestedLoc, extractType, loopNest.loops.back().getRegionIterArg(0),
+  sliceOffsets = SmallVector<OpFoldResult>(outputRank, zero);
+  sliceSizes = SmallVector<OpFoldResult>(outputRank, one);
+  sliceStrides = SmallVector<OpFoldResult>(outputRank, one);
+  if (contiguousAlongB) {
+    sliceSizes.front() = innerInputTileSize;
+    for (auto [idx, iv] : llvm::enumerate(ivs)) {
+      sliceOffsets[++idx] = iv;
+    }
+  } else {
+    sliceSizes.back() = innerInputTileSize;
+    for (auto [idx, iv] : llvm::enumerate(ivs)) {
+      sliceOffsets[idx] = iv;
+    }
+  }
+
+  Value inputForInsert;
+  // If the im2col input tensor has the batch dim at last and can be vectorized,
+  // transpose the dimensions to match the output order (Batch, M, K).
```

Comment:
nit: Maybe we can make this about the tiling instead of the vectorization, to follow Hanhan's suggestions.
```suggestion
  // If the batch dimension is untiled in the loop nest, transpose the dimensions
  // to match the output order (Batch, M, K).
```